http-clients/src/main/java/org/triplea/http/client/web/socket/WebSocketConnection.java

```diff
@@ -95,53 +109,71 @@ void close() {
   }
 
   /**
-   * Initiates a websocket connection. Must be called before {@link #sendMessage(String)}. This
-   * method blocks until the connection has either successfully been established, or the connection
-   * failed.
+   * Initiates a non-blocking websocket connection.
    *
-   * @throws CouldNotConnect If the connection fails
+   * @param errorHandler Invoked if there is a failure to connect.
+   * @throws IllegalStateException Thrown if connection is already open (eg: connect called twice).
+   * @throws IllegalStateException Thrown if connection has been closed (ie: 'close()' was called)
    */
-  void connect() {
+  // Suppression: Ignoring future return values ignores any exceptions thrown by the code
+  // that completes the future. In this case we ignore the returned future from 'whenComplete'
+  // to avoid blocking. We would not expect the ping sender start or error handler to throw
+  // exceptions.
+  @SuppressWarnings("FutureReturnValueIgnored")
+  void connect(final Consumer<String> errorHandler) {
     Preconditions.checkState(!client.isOpen());
     Preconditions.checkState(!closed);
-    try {
-      client.connect();
-      pingSender.start();
-    } catch (final Exception e) {
-      pingSender.cancel();
-      throw new CouldNotConnect(serverUri);
-    }
+
+    connectAsync()
+        .whenComplete(
+            (connected, throwable) -> {
+              if (connected && throwable == null) {
+                pingSender.start();
+              } else {
+                errorHandler.accept("Failed to connect to: " + serverUri);
+              }
+            });
+  }
+
+  // Suppression: Ignoring future return values ignores any exceptions thrown by the code
+  // that completes the future. Connect blocking does not throw any exceptions, there are
+  // no exceptions to ignore, we can ignore the future returned by 'submit'.
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private CompletableFuture<Boolean> connectAsync() {
+    final CompletableFuture<Boolean> completableFuture = new CompletableFuture<>();
+    // execute the connection attempt
+    Executors.newSingleThreadExecutor()
+        .submit(
+            () -> {
+              boolean connected;
+              try {
+                connected =
+                    client.connectBlocking(DEFAULT_CONNECT_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+              } catch (final InterruptedException ignored) {
+                connected = false;
+              }
+              return completableFuture.complete(connected);
```

Comment:
Any chance `WebSocketClient#connectBlocking` throws any other exception apart from `InterruptedException`?
I'd feel safer if we would add a `catch(Exception) {}` block that rejects the future, to prevent some user of this API to wait indefinitely

Reply:
`WebSocketConnection` is package protected, it's used by `GenericWebsocketConnection`. Thus the API usage is controlled. I would not want/expect for the visibility of `WebsocketConnection` to be raised, it really is meant to be used through `GenericWebsocketConnection`.

Here is the code behind `connect`, please note that it already catches exceptions very aggressively (and hence we do not need to do it, would be redundant; I originally had such a catch block and after reviewing the implementation code noted that it was not necessary):
```
	public void run() {
		InputStream istream;
		try {
			boolean isNewSocket = false;
			if (socketFactory != null) {
				socket = socketFactory.createSocket();
			} else if( socket == null ) {
				socket = new Socket( proxy );
				isNewSocket = true;

			} else if( socket.isClosed() ) {
				throw new IOException();
			}

			socket.setTcpNoDelay( isTcpNoDelay() );
			socket.setReuseAddress( isReuseAddr() );

			if( !socket.isBound() ) {
				socket.connect( new InetSocketAddress( uri.getHost(), getPort() ), connectTimeout );
			}

			// if the socket is set by others we don't apply any TLS wrapper
			if (isNewSocket && "wss".equals( uri.getScheme())) {

				SSLContext sslContext = SSLContext.getInstance("TLSv1.2");
				sslContext.init(null, null, null);
				SSLSocketFactory factory = sslContext.getSocketFactory();
				socket = factory.createSocket(socket, uri.getHost(), getPort(), true);
			}

			istream = socket.getInputStream();
			ostream = socket.getOutputStream();

			sendHandshake();
		} catch ( /*IOException | SecurityException | UnresolvedAddressException | InvalidHandshakeException | ClosedByInterruptException | SocketTimeoutException */Exception e ) {
			onWebsocketError( engine, e );
			engine.closeConnection( CloseFrame.NEVER_CONNECTED, e.getMessage() );
			return;
		}

		writeThread = new Thread( new WebsocketWriteThread(this) );
		writeThread.start();

		byte[] rawbuffer = new byte[ WebSocketImpl.RCVBUF ];
		int readBytes;

		try {
			while ( !isClosing() && !isClosed() && ( readBytes = istream.read( rawbuffer ) ) != -1 ) {
				engine.decode( ByteBuffer.wrap( rawbuffer, 0, readBytes ) );
			}
			engine.eot();
		} catch ( IOException e ) {
			handleIOException(e);
		} catch ( RuntimeException e ) {
			// this catch case covers internal errors only and indicates a bug in this websocket implementation
			onError( e );
			engine.closeConnection( CloseFrame.ABNORMAL_CLOSE, e.getMessage() );
		}
		connectReadThread = null;
	}
```